mtproto/session_store.go

```diff
@@ -47,15 +51,23 @@ func (s *SessFileStore) Save(sess *SessionInfo) (err error) {
 
 func (s *SessFileStore) Load(sess *SessionInfo) error {
 	f, err := os.Open(s.FPath)
-	if errors.Is(err, fs.ErrNotExist) {
+	if !s.RequirePresent && errors.Is(err, fs.ErrNotExist) {
 		return ErrNoSessionData.Here()
 	}
 	if err != nil {
-		return merry.Wrap(err)
+		if !s.RequireReadable {
```

Comment:
That way it will break the backward compatibility: store is currently created as `SessFileStore{"path"}` so `RequireReadable` will be false by default. But readability currently is required. So the flag should be inverted (`s.IgnoreUnreadable` / `s.SkipUnreadable` / ...)

Reply:
Don't think that handling this edgecase differently will break anything.

Reply:
Maaaybe. But it will also start skipping (by default) some errors which normally should not be suppressed. If session file has become corrupted/unreadable it's better not to try to proceed and overwrite the file, but stop and inform the user (he may want to recover that file from a dying file system for example).

Reply:
Program shouldn't refuse to work in such circumstances either. Maybe, a warning should be issued, if file is corrupted or unreadable? (And `s.DisableWarnings` flag added?)

Reply:
`ScanfAuthDataProvider` doesn't seem to overwrite anything, if user doesn't input data. And if user decides to use a custom authentication provider, one will take care of handling errors properly, if needed. For most programs suggested default behaviour seems reasonable.

Reply:
> Program shouldn't refuse to work in such circumstances either.

Program *should* refuse to work if it's database got corrupted.

> `ScanfAuthDataProvider` doesn't seem to overwrite anything

Yes, because it doesn't know anything about session, it just returns some strings. So it can not take care of session errors.

Also, `ScanfAuthDataProvider`'s prompt will not prevent session overwrite. After your change and in case of corrupted session file, `InitSession()` will create new random session id (in memory), then `initConection()` will obtain auth key from TG for that session ID (`makeAuthKey()`) and *save that session to a file*. And only then that session will be used for `Auth()` with phone number prompt.

Session may be even used without `Auth()` for some methods ([help.getNearestDc](https://core.telegram.org/method/help.getNearestDc) for example).

Reply:
> Program should refuse to work if it's database got corrupted.

It is not a *database*, the *database* is Telegram, it is more of a authentication cookie, which is easily renewable.

Reply:
> Also, `ScanfAuthDataProvider`'s prompt will not prevent session overwrite. After your change and in case of corrupted session file, `InitSession()` will create new random session id (in memory), then `initConection()` will obtain auth key from TG for that session ID (`makeAuthKey()`) and _save that session to a file_. And only then that session will be used for `Auth()` with phone number prompt.

It may be reasonable to not to write session to file, till it can be used to log in. Are there reasons not to do so?

Reply:
> it is more of a authentication cookie

The cookie store is also a database :-) Very small and simple.

> cookie, which is easily renewable

Not always. The account could be registered via a one-time phone number, the cloud password may have been lost, etc.

> It may be reasonable to not to write session to file, till it can be used to log in. Are there reasons not to do so?

As I said, some TG API methods do not require authentication, so someone may use tgclient without being logging in. Session file will never be written and auth keys will need to be generated on every run. This requires factorization of a large number and is relatively slow.

Reply:
> The cookie store is also a database :-) Very small and simple.

Websites don't refuse to work, if cookie is damaged

Reply:
Nor should programs using the library, I guess

Reply:
> Websites don't refuse to work, if cookie is damaged

But can it be actually *damaged*? If cookie *value* is damaged (string with a valid number was written but then became invalid for example), the site will likely break, because user is not supposed to change that value by hand and browser should guarantee cookie store integrity.

And if the whole browser *session storage* is damaged, it (I suppose) will be just re-created and user will loose his cookies. Because browser user is unlikely to know what to do with the broken cookie file. But tgclient user is likely know.